File: CenterofBlob/single_blob.cpp
// OpenCV core defines Mat, Point, Scalar, Moments, and countNonZero.
#include <opencv2/core.hpp>
// The version header exposes CV_VERSION and CV_VERSION_MAJOR.
#include <opencv2/core/version.hpp>
// highgui supplies the optional imshow, waitKey, and destroyAllWindows calls.
#include <opencv2/highgui.hpp>
// imgcodecs supplies imread and imwrite for the tracked PNG and generated output.
#include <opencv2/imgcodecs.hpp>
// imgproc supplies color conversion, thresholding, circles, and text rendering.
#include <opencv2/imgproc.hpp>

// OpenCV 5 moved moments and related shape functions into geometry.
#if CV_VERSION_MAJOR >= 5
#include <opencv2/geometry.hpp>
#endif

// cmath supplies std::abs for the zero-area guard.
#include <cmath>
// cstdlib supplies EXIT_SUCCESS and EXIT_FAILURE.
#include <cstdlib>
// exception supplies the common base type used by the CLI error boundary.
#include <exception>
// filesystem provides portable input and output paths in C++17.
#include <filesystem>
// iostream provides stable result output and clear error messages.
#include <iostream>
// optional represents the absence of an output destination without sentinels.
#include <optional>
// stdexcept provides invalid_argument and runtime_error.
#include <stdexcept>
// string stores command-line values before they are parsed.
#include <string>
// system_error lets filesystem operations report failures without ambiguity.
#include <system_error>

// Direct compiler invocations can still run by explicitly passing an input path.
#ifndef CENTER_OF_BLOB_SOURCE_DIR
#define CENTER_OF_BLOB_SOURCE_DIR "."
#endif

namespace {

// Use a short namespace alias for the many C++17 path operations below.
namespace fs = std::filesystem;

// CMake replaces this directory at compile time so the default works from any cwd.
const fs::path kDefaultInputPath =
    fs::path(CENTER_OF_BLOB_SOURCE_DIR) / "circle.png";
// Directory-based output always receives one predictable filename.
constexpr const char *kDefaultOutputName = "single_blob_result.png";
// These are the stable measurements of the tracked tutorial asset.
constexpr int kExpectedCenterX = 112;
constexpr int kExpectedCenterY = 112;
constexpr int kExpectedForegroundPixels = 35'272;

// Options records only parsed user intent and contains no OpenCV state.
struct Options {
  // The bundled asset is the default; --input or a positional path overrides it.
  fs::path image_path = kDefaultInputPath;
  // Track explicit input selection to reject two conflicting input spellings.
  bool input_was_set = false;
  // Threshold 127 separates the black circle from the white background.
  int threshold_value = 127;
  // The bundled example contains a dark object on a light background.
  bool foreground_dark = true;
  // --output writes to an exact caller-selected filename.
  std::optional<fs::path> output_path;
  // --output-dir writes kDefaultOutputName inside a caller-selected directory.
  std::optional<fs::path> output_directory;
  // Display is enabled for interactive teaching and disabled by --no-display.
  bool display = true;
  // --validate checks immutable sample metrics and emits a success marker.
  bool validate = false;
};

// Print both preferred switches and the legacy positional form.
void print_usage(const char *program) {
  // Keep the usage on one logical line so shell users can copy it easily.
  std::cout
      << "Usage: " << program
      << " [IMAGE | --input IMAGE] [--threshold 0..255]"
         " [--foreground dark|light] [--output FILE | --output-dir DIR]"
         " [--no-display] [--validate]\n";
}

// Parse the threshold strictly so values such as "12x" are not truncated.
int parse_threshold(const std::string &value) {
  // stoi reports how many input characters participated in the conversion.
  std::size_t parsed = 0;
  // Initialize before the try block to keep the range check straightforward.
  int result = 0;
  try {
    // Convert the complete decimal representation into an integer.
    result = std::stoi(value, &parsed);
  } catch (const std::exception &) {
    // Translate library-specific conversion failures into a stable CLI message.
    throw std::invalid_argument("--threshold expects an integer");
  }
  // Reject trailing characters and values outside an 8-bit intensity range.
  if (parsed != value.size() || result < 0 || result > 255) {
    throw std::invalid_argument("--threshold must be between 0 and 255");
  }
  // Return only a completely parsed, range-checked value.
  return result;
}

// Parse all supported spellings without reading files or invoking OpenCV.
Options parse_arguments(int argc, char **argv) {
  // Start with bundled-input and tutorial-friendly defaults.
  Options options;
  // Examine every token after the executable name exactly once.
  for (int index = 1; index < argc; ++index) {
    // Copy the current token so comparisons do not depend on raw pointers.
    const std::string argument = argv[index];
    // --help is a successful terminal action rather than an error.
    if (argument == "--help" || argument == "-h") {
      print_usage(argv[0]);
      std::exit(EXIT_SUCCESS);
    }
    // Automated tests must never wait for a window-system event.
    if (argument == "--no-display") {
      options.display = false;
      continue;
    }
    // Validation is independent of whether an output image is requested.
    if (argument == "--validate") {
      options.validate = true;
      continue;
    }
    // These switches each consume exactly one following token.
    if (argument == "--input" || argument == "-i" ||
        argument == "--ipimage" || argument == "--threshold" ||
        argument == "--foreground" || argument == "--output" ||
        argument == "--output-dir") {
      // Fail clearly rather than reading beyond argv when a value is absent.
      if (index + 1 >= argc) {
        throw std::invalid_argument("Missing value for " + argument);
      }
      // Advance once so the outer loop resumes after the consumed value.
      const std::string value = argv[++index];
      // All input aliases select the same image-path field.
      if (argument == "--input" || argument == "-i" ||
          argument == "--ipimage") {
        // Two inputs are ambiguous even when they happen to name the same file.
        if (options.input_was_set) {
          throw std::invalid_argument("Only one input image may be specified");
        }
        options.image_path = value;
        options.input_was_set = true;
      } else if (argument == "--threshold") {
        // Apply strict integer and range validation immediately.
        options.threshold_value = parse_threshold(value);
      } else if (argument == "--foreground") {
        // Only the two documented polarity words have defined semantics.
        if (value != "dark" && value != "light") {
          throw std::invalid_argument(
              "--foreground must be 'dark' or 'light'");
        }
        options.foreground_dark = value == "dark";
      } else if (argument == "--output") {
        // The two output modes would otherwise target competing filenames.
        if (options.output_directory.has_value()) {
          throw std::invalid_argument(
              "--output and --output-dir cannot be used together");
        }
        options.output_path = value;
      } else {
        // The remaining recognized value switch is --output-dir.
        if (options.output_path.has_value()) {
          throw std::invalid_argument(
              "--output and --output-dir cannot be used together");
        }
        options.output_directory = value;
      }
      continue;
    }
    // Any other dash-prefixed token is probably a misspelled option.
    if (!argument.empty() && argument.front() == '-') {
      throw std::invalid_argument("Unknown option: " + argument);
    }
    // Preserve the original tutorial's positional input-image syntax.
    if (options.input_was_set) {
      throw std::invalid_argument("Only one input image may be specified");
    }
    options.image_path = argument;
    options.input_was_set = true;
  }
  // Omitting an input intentionally leaves the bundled default in place.
  return options;
}

// Convert a BGR image into a mask whose intended foreground is nonzero.
cv::Mat create_binary_mask(const cv::Mat &image, int threshold_value,
                           bool foreground_dark) {
  // Convert to one channel because thresholding color channels independently
  // would make image moments difficult to interpret.
  cv::Mat gray;
  cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
  // Allocate the result through OpenCV so its size and type follow gray.
  cv::Mat binary_mask;
  // Invert dark objects so moments measure the object instead of its background.
  const int threshold_type =
      foreground_dark ? cv::THRESH_BINARY_INV : cv::THRESH_BINARY;
  // Produce a strict 0/255 mask that is portable across OpenCV 4.14 and 5.
  cv::threshold(gray, binary_mask, threshold_value, 255, threshold_type);
  // Return the mask for both moment calculation and regression validation.
  return binary_mask;
}

// Compute the area-normalized first-order moment of the foreground mask.
cv::Point find_centroid(const cv::Mat &binary_mask) {
  // binaryImage=true gives every nonzero mask pixel equal weight.
  const cv::Moments image_moments = cv::moments(binary_mask, true);
  // m00 is zero when thresholding found no foreground pixels.
  if (std::abs(image_moments.m00) <= 1e-12) {
    throw std::runtime_error(
        "No foreground pixels were found in the binary mask");
  }
  // cvRound matches OpenCV's conventional conversion from subpixel centers.
  return {cvRound(image_moments.m10 / image_moments.m00),
          cvRound(image_moments.m01 / image_moments.m00)};
}

// Check all stable invariants of the repository's bundled circle image.
void validate_bundled_result(const fs::path &input_path, const cv::Mat &image,
                             const cv::Mat &binary_mask,
                             const cv::Point &centroid) {
  // equivalent compares filesystem identity and handles relative input spellings.
  std::error_code equivalence_error;
  const bool is_bundled =
      fs::equivalent(input_path, kDefaultInputPath, equivalence_error);
  // Never print a success marker for an arbitrary custom image.
  if (equivalence_error || !is_bundled) {
    throw std::runtime_error(
        "--validate requires the bundled circle.png input");
  }
  // Image dimensions and channel count verify successful decoding.
  if (image.rows != 225 || image.cols != 225 || image.channels() != 3) {
    throw std::runtime_error("Unexpected circle.png dimensions or channels");
  }
  // The centroid is the tutorial's central numerical result.
  if (centroid.x != kExpectedCenterX || centroid.y != kExpectedCenterY) {
    throw std::runtime_error("Bundled circle centroid did not match (112, 112)");
  }
  // The centered background shares this centroid, so also verify polarity using
  // the exact number of foreground pixels in the thresholded lossless PNG.
  if (cv::countNonZero(binary_mask) != kExpectedForegroundPixels) {
    throw std::runtime_error(
        "Bundled circle foreground-pixel count did not match 35272");
  }
}

// Resolve either output mode and create only its narrowly scoped parent directory.
std::optional<fs::path> resolve_output_path(const Options &options) {
  // An explicit output filename may include a new nested parent directory.
  if (options.output_path.has_value()) {
    // Avoid create_directories("") for a filename in the current directory.
    const fs::path parent = options.output_path->parent_path();
    if (!parent.empty()) {
      std::error_code create_error;
      fs::create_directories(parent, create_error);
      if (create_error) {
        throw std::runtime_error("Could not create output directory '" +
                                 parent.string() + "': " +
                                 create_error.message());
      }
    }
    return options.output_path;
  }
  // No output option means the example only prints and optionally displays.
  if (!options.output_directory.has_value()) {
    return std::nullopt;
  }
  // Create the requested directory before appending the deterministic filename.
  std::error_code create_error;
  fs::create_directories(*options.output_directory, create_error);
  if (create_error) {
    throw std::runtime_error("Could not create output directory '" +
                             options.output_directory->string() + "': " +
                             create_error.message());
  }
  return *options.output_directory / kDefaultOutputName;
}

// Encode the annotation and optionally prove the generated file is readable.
void write_annotated_image(const fs::path &output_path,
                           const cv::Mat &annotated, bool validate) {
  // imwrite returns false for encoder/write failures that do not throw.
  if (!cv::imwrite(output_path.string(), annotated)) {
    throw std::runtime_error("Could not write output image '" +
                             output_path.string() + "'");
  }
  // Validation checks the actual file rather than trusting only the return value.
  if (validate) {
    const cv::Mat decoded =
        cv::imread(output_path.string(), cv::IMREAD_COLOR);
    if (decoded.empty() || decoded.size() != annotated.size() ||
        decoded.type() != annotated.type()) {
      throw std::runtime_error("Could not verify output image '" +
                               output_path.string() + "'");
    }
  }
}

} // namespace

// The executable boundary converts all expected failures into a nonzero exit.
int main(int argc, char **argv) {
  try {
    // Parse CLI intent before performing image or filesystem operations.
    const Options options = parse_arguments(argc, argv);
    // Decode in color because annotation uses colored text and shapes.
    const cv::Mat image =
        cv::imread(options.image_path.string(), cv::IMREAD_COLOR);
    // A missing or unsupported input should not reach cvtColor.
    if (image.empty()) {
      std::cerr << "Error: could not read input image '"
                << options.image_path.string() << "'\n";
      return EXIT_FAILURE;
    }

    // Threshold with an explicit foreground polarity.
    const cv::Mat binary_mask = create_binary_mask(
        image, options.threshold_value, options.foreground_dark);
    // Measure the intended nonzero blob rather than the white background.
    const cv::Point centroid = find_centroid(binary_mask);
    // Clone before drawing so the decoded input remains unchanged.
    cv::Mat annotated = image.clone();
    // Mark the measured center with a filled white circle.
    cv::circle(annotated, centroid, 5, cv::Scalar(255, 255, 255),
               cv::FILLED);
    // Label the marker with APIs shared by OpenCV 4.14 and OpenCV 5.
    cv::putText(annotated, "centroid", centroid - cv::Point(25, 25),
                cv::FONT_HERSHEY_SIMPLEX, 0.5,
                cv::Scalar(255, 255, 255), 2, cv::LINE_8);

    // Resolve output before validation so invalid destinations fail predictably.
    const std::optional<fs::path> output_path =
        resolve_output_path(options);
    // Verify the complete sample contract when regression mode is enabled.
    if (options.validate) {
      validate_bundled_result(options.image_path, image, binary_mask, centroid);
    }
    // Save only when the caller requested an output filename or directory.
    if (output_path.has_value()) {
      write_annotated_image(*output_path, annotated, options.validate);
    }

    // Print the stable primary metric in the same form as the Python example.
    std::cout << "Centroid: (" << centroid.x << ", " << centroid.y << ")\n";
    // Record the linked OpenCV version in every test log.
    std::cout << "OpenCV version: " << CV_VERSION << '\n';
    // Emit a marker only after metrics and optional output verification pass.
    if (options.validate) {
      std::cout << "VALIDATION PASSED: single_blob\n";
    }
    // Keep display available for readers while allowing fully headless tests.
    if (options.display) {
      cv::imshow("Image with center", annotated);
      cv::waitKey(0);
      cv::destroyAllWindows();
    }
  } catch (const std::exception &error) {
    // Present OpenCV, parsing, validation, and filesystem errors uniformly.
    std::cerr << "Error: " << error.what() << '\n';
    print_usage(argv[0]);
    return EXIT_FAILURE;
  }
  // Reaching this line means processing and every requested check succeeded.
  return EXIT_SUCCESS;
}

File: CenterofBlob/single_blob.py
#!/usr/bin/env python3
"""Find and annotate the centroid of one foreground blob."""

from __future__ import annotations

# argparse provides a documented command-line interface for the tutorial.
import argparse
# sys lets errors use stderr while normal results remain on stdout.
import sys
# Path makes bundled assets independent of the caller's working directory.
from pathlib import Path
# Optional keeps type annotations compatible with the documented Python 3.9.
from typing import Optional

# OpenCV supplies image I/O, thresholding, moments, drawing, and display APIs.
import cv2


# Resolve the example directory once so the default image works from any cwd.
SCRIPT_DIRECTORY = Path(__file__).resolve().parent
# Use the tutorial's tracked circle image when the caller omits --input.
DEFAULT_INPUT_PATH = SCRIPT_DIRECTORY / "circle.png"
# Give --output-dir a deterministic filename that tests can assert exactly.
DEFAULT_OUTPUT_NAME = "single_blob_result.png"
# The tracked image has a stable foreground centroid under the documented mask.
EXPECTED_CENTROID = (112, 112)
# This count catches accidental foreground/background inversions.
EXPECTED_FOREGROUND_PIXELS = 35_272


def create_binary_mask(
    image, threshold_value: int = 127, foreground: str = "dark"
):
    """Convert a BGR image to a mask whose foreground pixels are white."""
    # Reject an absent or zero-sized image before cvtColor emits a cryptic error.
    if image is None or image.size == 0:
        raise ValueError("The input image is empty")
    # OpenCV thresholds use the inclusive 8-bit intensity range.
    if not 0 <= threshold_value <= 255:
        raise ValueError("threshold_value must be between 0 and 255")
    # Keep the polarity choice explicit instead of silently guessing.
    if foreground not in {"dark", "light"}:
        raise ValueError("foreground must be 'dark' or 'light'")

    # Moments need a one-channel mask, so discard color after loading the image.
    gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    # Dark objects must be inverted so the intended blob, not its background,
    # becomes the nonzero foreground measured by cv2.moments.
    threshold_type = (
        cv2.THRESH_BINARY_INV if foreground == "dark" else cv2.THRESH_BINARY
    )
    # Convert the grayscale image into an 8-bit mask containing only 0 and 255.
    _, binary_mask = cv2.threshold(
        gray_image, threshold_value, 255, threshold_type
    )
    # Return the mask as well as using it internally so tests can inspect it.
    return binary_mask


def find_centroid(binary_mask) -> tuple[int, int]:
    """Return the centroid of the nonzero pixels in a binary mask."""
    # binaryImage=True treats every nonzero pixel equally, independent of value.
    moments = cv2.moments(binary_mask, binaryImage=True)
    # m00 is the foreground area; division by zero means no blob was detected.
    if abs(moments["m00"]) <= 1e-12:
        raise ValueError("No foreground pixels were found in the binary mask")

    # Normalize the first-order x moment by area to obtain the horizontal center.
    center_x = int(round(moments["m10"] / moments["m00"]))
    # Normalize the first-order y moment by area to obtain the vertical center.
    center_y = int(round(moments["m01"] / moments["m00"]))
    # A small immutable tuple is convenient for both drawing and validation.
    return center_x, center_y


def annotate_centroid(image, centroid: tuple[int, int]):
    """Return a copy of the image with its centroid marked."""
    # Never paint over the caller's input; preserving it simplifies comparisons.
    annotated = image.copy()
    # Unpack the coordinates for the readable text-offset calculation below.
    center_x, center_y = centroid
    # Draw a filled white dot directly over the measured center.
    cv2.circle(annotated, centroid, 5, (255, 255, 255), cv2.FILLED)
    # Label the point using constants available in both OpenCV 4.14 and 5.
    cv2.putText(
        annotated,
        "centroid",
        (center_x - 25, center_y - 25),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.5,
        (255, 255, 255),
        2,
        cv2.LINE_8,
    )
    # Return the annotated copy so GUI display remains optional.
    return annotated


def process_image(
    image, threshold_value: int = 127, foreground: str = "dark"
):
    """Return ``(annotated_image, centroid, binary_mask)``."""
    # Build a foreground mask with the selected threshold and polarity.
    binary_mask = create_binary_mask(image, threshold_value, foreground)
    # Compute the center from mask pixels rather than from the white background.
    centroid = find_centroid(binary_mask)
    # Keep all useful intermediate results available to tests and other callers.
    return annotate_centroid(image, centroid), centroid, binary_mask


def validate_bundled_result(
    input_path: Path, image, binary_mask, centroid: tuple[int, int]
) -> None:
    """Raise ValueError unless the bundled-image regression checks pass."""
    # Validation values apply only to the immutable tutorial asset.
    if input_path.resolve() != DEFAULT_INPUT_PATH.resolve():
        raise ValueError("--validate requires the bundled circle.png input")
    # Check the decoded dimensions before trusting any derived measurement.
    if image.shape != (225, 225, 3):
        raise ValueError(f"Unexpected circle.png shape: {image.shape}")
    # A centroid mismatch would change the tutorial's central numerical result.
    if centroid != EXPECTED_CENTROID:
        raise ValueError(
            f"Expected centroid {EXPECTED_CENTROID}, received {centroid}"
        )
    # Pixel count also detects a threshold-polarity regression that the centered
    # source image could otherwise hide because its background has the same center.
    foreground_pixels = cv2.countNonZero(binary_mask)
    if foreground_pixels != EXPECTED_FOREGROUND_PIXELS:
        raise ValueError(
            "Expected "
            f"{EXPECTED_FOREGROUND_PIXELS} foreground pixels, "
            f"received {foreground_pixels}"
        )


def resolve_output_path(
    output: Optional[Path], output_directory: Optional[Path]
) -> Optional[Path]:
    """Create the requested directory and return the final output path."""
    # An explicit filename is useful for backward-compatible one-off commands.
    if output is not None:
        # Create its parent as well, so nested paths behave predictably.
        output.parent.mkdir(parents=True, exist_ok=True)
        return output
    # With no output option, the example only prints and optionally displays.
    if output_directory is None:
        return None
    # CI uses a fresh directory and expects one exact, deterministic filename.
    output_directory.mkdir(parents=True, exist_ok=True)
    return output_directory / DEFAULT_OUTPUT_NAME


def write_annotated_image(output_path: Path, annotated, validate: bool) -> None:
    """Write an annotation and optionally prove OpenCV can read it back."""
    # imwrite returns False for many failures instead of raising an exception.
    try:
        written = cv2.imwrite(str(output_path), annotated)
    except cv2.error as error:
        raise OSError(f"could not write output image '{output_path}': {error}") from error
    # Convert the boolean failure into a clear tutorial-facing message.
    if not written:
        raise OSError(f"could not write output image '{output_path}'")
    # Validation checks the actual encoded file, not just imwrite's return value.
    if validate:
        decoded = cv2.imread(str(output_path), cv2.IMREAD_COLOR)
        if decoded is None or decoded.shape != annotated.shape:
            raise OSError(f"could not verify output image '{output_path}'")


def build_argument_parser() -> argparse.ArgumentParser:
    """Describe the command-line interface in one testable function."""
    # Defaults in ArgumentParser automatically appear in --help output.
    parser = argparse.ArgumentParser(description="Find the center of a single blob.")
    # --ipimage keeps old article commands working; --input is the preferred name.
    parser.add_argument(
        "-i",
        "--input",
        "--ipimage",
        dest="input_path",
        type=Path,
        default=DEFAULT_INPUT_PATH,
        help="input image (default: bundled circle.png)",
    )
    # Expose the tutorial threshold so other high-contrast images can be tested.
    parser.add_argument(
        "--threshold",
        type=int,
        default=127,
        help="grayscale threshold from 0 to 255 (default: 127)",
    )
    # The default matches the bundled dark circle on a light background.
    parser.add_argument(
        "--foreground",
        choices=("dark", "light"),
        default="dark",
        help="whether the blob is darker or lighter than the background",
    )
    # Only one output naming mode may be selected for an invocation.
    output_group = parser.add_mutually_exclusive_group()
    # Preserve the draft migration's explicit output-file interface.
    output_group.add_argument(
        "--output",
        type=Path,
        help="optional path for the annotated image",
    )
    # Add the standard migration interface used by repeatable test runs.
    output_group.add_argument(
        "--output-dir",
        type=Path,
        help=f"write {DEFAULT_OUTPUT_NAME} in this directory",
    )
    # Headless operation prevents imshow/waitKey from blocking automated tests.
    parser.add_argument(
        "--no-display", action="store_true", help="do not open a GUI window"
    )
    # Validation checks the known metrics of the bundled image.
    parser.add_argument(
        "--validate",
        action="store_true",
        help="validate the bundled image and print a success marker",
    )
    # Return the parser so tests can inspect parsing without running OpenCV.
    return parser


def run(args: argparse.Namespace) -> int:
    """Execute one CLI invocation and return a process-style status code."""
    # Decode in color because the annotation is drawn on a three-channel image.
    image = cv2.imread(str(args.input_path), cv2.IMREAD_COLOR)
    # A missing or unsupported image should fail before any OpenCV operation.
    if image is None:
        print(
            f"Error: could not read input image '{args.input_path}'",
            file=sys.stderr,
        )
        return 1

    try:
        # Exercise the same callable path used by unit tests.
        annotated, centroid, binary_mask = process_image(
            image, args.threshold, args.foreground
        )
        # Resolve and create output paths before reporting successful completion.
        output_path = resolve_output_path(args.output, args.output_dir)
        # Validate stable tutorial metrics when the caller requests regression mode.
        if args.validate:
            validate_bundled_result(
                args.input_path, image, binary_mask, centroid
            )
        # Encode the annotation only when an output option was supplied.
        if output_path is not None:
            write_annotated_image(output_path, annotated, args.validate)
    # ValueError covers validation/input semantics; OSError covers output paths.
    except (ValueError, OSError) as error:
        print(f"Error: {error}", file=sys.stderr)
        return 1

    # Print the primary result in a stable form shared with the C++ example.
    print(f"Centroid: ({centroid[0]}, {centroid[1]})")
    # Emit the exact runtime version to make matrix evidence unambiguous.
    print(f"OpenCV version: {cv2.__version__}")
    # A machine-readable marker lets CTest-style harnesses reject silent success.
    if args.validate:
        print("VALIDATION PASSED: single_blob")

    # Keep visualization opt-in for CI but on by default for tutorial use.
    if not args.no_display:
        cv2.imshow("Image with center", annotated)
        cv2.waitKey(0)
        cv2.destroyAllWindows()
    # Zero communicates that processing, optional output, and validation succeeded.
    return 0


def main() -> int:
    """Parse arguments and run the example."""
    # Construct the parser inside main so importing the module has no side effects.
    parser = build_argument_parser()
    # Parse the process command line using argparse's standard error reporting.
    args = parser.parse_args()
    # Keep threshold validation explicit because argparse only converts the type.
    if not 0 <= args.threshold <= 255:
        parser.error("--threshold must be between 0 and 255")
    # Hand the validated namespace to a directly testable execution function.
    return run(args)


# Importing this file exposes its functions without opening windows or reading data.
if __name__ == "__main__":
    raise SystemExit(main())
